### PR TITLE
Refactor tls-client example to improve readability

### PR DESCRIPTION
**Pull request by andresag01**
This patch completely refactors the tls-client example. The following
is a list of significant changes:
    * Split the code into .h and .cpp files.
    * Move the class HelloHTTPS to its own file.
    * Remove all unused variables and redundant buffers.
    * Split functionality into smaller, more manageable functions.
    * Improve error reporting and logging.

### CODE REVIEW

**Comment by andresag01** [comment]
@yanesca @mazimkhan @sbutcher-arm: Could you please review?

The CI ran as expected and reported the same known failures:
* NUMAKER_PFM_NUC472 fails with all compilers
* UBLOX_EVK_ODIN_W2 fails with IAR (this fails because the mbed OS patch that fixes it has not yet been released).

File: tls-client/main.cpp [hunk 1]
```diff
@@ -19,417 +19,72 @@
...
-        /* Extract basename from file */
-        for(p = basename = file; *p != '\0'; p++) {
-            if(*p == '/' || *p == '\\') {
-                basename = p + 1;
-            }
-        }
-
-        mbedtls_printf("%s:%04d: |%d| %s", basename, line, level, str);
+        mbedtls_printf( "Failed to allocate HelloHttpsClient object\r\n" );
+        goto exit;
```
Comment (mazimkhan) [inline tls-client/main.cpp]: Just for 2 ```if``` blocks I would not use ```goto```. Instead a nested ```if``` would be more clear.

File: tls-client/main.cpp [hunk 2]
```diff
@@ -19,417 +19,72 @@
...
-        if(NSAPI_ERROR_WOULD_BLOCK == size){
-            return MBEDTLS_ERR_SSL_WANT_WRITE;
-        }else if(size < 0){
-            mbedtls_printf("Socket send error %d\r\n", size);
-            return -1;
-        }else{
-            return size;
-        }
+exit:
+    if( exit_code == MBEDTLS_EXIT_SUCCESS )
```
Comment (mazimkhan) [inline tls-client/main.cpp]: Why do we have to instrument printing ```DONE``` or ```FAIL```. It looks like a test case output. Is it not clear that an example failed when it does not reach ```DONE```.

File: tls-client/HelloHttpsClient.h [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    const char *server_name;
+    /**
+     * The port number to use in the connection
+     */
+    const uint16_t server_port;
+
+    /**
+     * A generic buffer used to hold debug or HTTP request/response strings
+     */
+    char gp_buf[GENERAL_PURPOSE_BUFFER_LENGTH];
```
Comment (mazimkhan) [inline tls-client/HelloHttpsClient.h]: No need to pre-allocate this buffer as part of this class. It can be allocated when needed. That is when ssl write is done. Before ssl write the ssl handshake API grows memory for certificate verification and key comparisons. Finally it frees the memory. A pre allocated buffer will keep the mem at that peak time.
It may help it to work on memory constrained platforms. Also a good practice.

File: tls-client/main.cpp (hunk 2 above)
Comment (andresag01) [inline tls-client/main.cpp]: It makes it easier to check from `htrun` whether the example fully worked (including the teardown logic).

File: tls-client/main.cpp (hunk 1 above)
Comment (andresag01) [inline tls-client/main.cpp]: I disagree. Throughout the example and in the mbed TLS library, this is the way we generally handle failures, and in my opinion it is very clear. It also reduces code duplication and helps maintenance as there is only one exit point for the function. I suppose that at the end of the day, this is just favourite coding practices.

File: tls-client/HelloHttpsClient.h (hunk 3 above)
Comment (andresag01) [inline tls-client/HelloHttpsClient.h]: Agreed, assuming that the handshake is the peak consumption point. However, I am not sure this will cause an issue as this is allocated in the heap. What do the other reviewers think?

File: tls-client/main.cpp (hunk 2 above)
Comment (mazimkhan) [inline tls-client/main.cpp]: ```htrun``` has log comparison feature for detecting pass or fail. It does not look for ```DONE``` or ```FAIL```.

File: tls-client/main.cpp (hunk 1 above)
Comment (mazimkhan) [inline tls-client/main.cpp]: It is not about what is already there in lib. ```goto``` is only help when there is functions long sequence of function calls that need different cleanups. Here there are only two. A nested if would be better. But it is just my opinion.

File: tls-client/main.cpp (hunk 2 above)
Comment (andresag01) [inline tls-client/main.cpp]: @mazimkhan I dont understand what you mean. I thought `htrun` only checks the serial output (just as the user would do), how can it tell what the return code of the function is?

File: tls-client/HelloHttpsClient.cpp [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+    delete[] buf;
+
+exit:
+    return( ret );
+}
+
+int HelloHttpsClient::run( void )
+{
+    int ret;
+    size_t req_len, req_offset, resp_offset;
```
Comment (mazimkhan) [inline tls-client/HelloHttpsClient.cpp]: C++ does not enforce declaring variables at top. They should be as close to the usage as possible. These could declared and assigned at https://github.com/ARMmbed/mbed-os-example-tls/pull/63/files#diff-39d7102cc3f72bf4a2240cc439f38d0fR388
This also avoids uninitialised use.

File: tls-client/HelloHttpsClient.cpp [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+    uint32_t flags;
+    bool resp_200, resp_hello;
```
Comment (mazimkhan) [inline tls-client/HelloHttpsClient.cpp]: same as above.

File: tls-client/HelloHttpsClient.cpp [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+    mbedtls_printf( "%s:%d: |%d| %s\r", basename, line, level, str );
+}
+
+int HelloHttpsClient::sslVerify( void *data, mbedtls_x509_crt *crt, int depth,
+                                 uint32_t *flags )
+{
+    (void)data;
+
+    int ret = -1;
+    char *buf = new char[GENERAL_PURPOSE_BUFFER_LENGTH];
```
Comment (mazimkhan) [inline tls-client/HelloHttpsClient.cpp]: Is there a reason we cannot use ```gp_buf```` here? It is fine if we decide to use a localised ```gp_buf```` at the other place.

File: tls-client/HelloHttpsClient.cpp [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+
+    return( ret );
+}
+
+void HelloHttpsClient::sslDebug( void *ctx, int level, const char *file,
+                                 int line, const char *str )
+{
+    (void)ctx;
+
+    const char *p, *basename;
```
Comment (mazimkhan) [inline tls-client/HelloHttpsClient.cpp]: Uninitialised variables. If below loop fails to find ```basename``` then we could supplying a garbage value to the print functions below.

File: tls-client/HelloHttpsClient.cpp [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+    /* Extract basename from file */
+    for( p = basename = file; *p != '\0'; p++ )
+    {
+        if( *p == '/' || *p == '\\' )
+        {
+            basename = p + 1;
+        }
+    }
+
+    mbedtls_printf( "%s:%d: |%d| %s\r", basename, line, level, str );
```
Comment (mazimkhan) [inline tls-client/HelloHttpsClient.cpp]: Check if ```basename``` is valid by initializing it to NULL and checking here.

**Comment by mazimkhan** [comment]
In general example code should be concise and quick to grasp. Scattering code should be avoided. Only take away for the users is the API usage.

**Comment by andresag01** [comment]
Closing as this PR was created against the development branch, which will be deleted.

**Comment by mazimkhan** [comment]
@andresag01 I think you were referring to this comment of mine https://github.com/ARMmbed/mbed-os-example-tls/pull/63#issuecomment-285700683
This comment is not against splitting a class but rather using a class for demonstrating use of a C library. The reason why I opposed it was that I believe an example should focus on the API that we want to demonstrate than adding extra things that can convolute the code and hence make it hard to follow by the intended users. In past there were some worst examples in BLE repo.
cc @Patater
